File: aitest_kit/console/agent_diagnostics.py
"""Bounded diagnostic projection independent of the live SSE replay window."""
from __future__ import annotations

from copy import deepcopy
from typing import Any, Iterable, Mapping


class DiagnosticProjection:
    def __init__(self, events: Iterable[Mapping[str, Any]] = ()) -> None:
        self._requests: dict[str, dict[str, Any]] = {}
        self._retry: dict[str, Any] | None = None
        self._message_id = ""
        for event in events:
            self.apply(str(event["type"]), event["payload"], str(event.get("correlation_id", "")))

    def apply(self, event_type: str, payload: Mapping[str, Any], correlation_id: str = "") -> None:
        if event_type == "request_diagnostic" and isinstance(payload.get("request_id"), str):
            self._requests[payload["request_id"]] = deepcopy(dict(payload))
            self._message_id = str(payload.get("message_id", ""))
            if payload.get("phase") == "preparing" and self._retry:
                self._retry["waiting"] = False
            while len(self._requests) > 20:
                del self._requests[next(iter(self._requests))]
        elif event_type == "agent_retry":
            self._retry = {**deepcopy(dict(payload)), "waiting": payload.get("phase") == "start"}
        elif event_type == "user_message":
            self._retry = None
            self._message_id = correlation_id

    def snapshot(self) -> dict[str, Any]:
        return deepcopy({"requests": list(self._requests.values()), "retry": self._retry, "message_id": self._message_id})

File: console_web/src/agent-diagnostics.ts
import type { AgentDiagnosticsSnapshot, AgentEvent, AgentRequestDiagnostic, AgentRetryDiagnostic } from './types'

export function projectDiagnostics(previous: AgentDiagnosticsSnapshot | undefined, event: AgentEvent): AgentDiagnosticsSnapshot {
  const state: AgentDiagnosticsSnapshot = previous ?? { requests: [], retry: null, message_id: '' }
  if (event.type === 'request_diagnostic') {
    const row = event.payload as unknown as AgentRequestDiagnostic
    const index = state.requests.findIndex((item) => item.request_id === row.request_id)
    if (index >= 0) state.requests[index] = row
    else state.requests.push(row)
    state.requests = state.requests.slice(-20)
    state.message_id = row.message_id
    if (row.phase === 'preparing' && state.retry) state.retry.waiting = false
  } else if (event.type === 'agent_retry') {
    state.retry = { ...event.payload, waiting: event.payload.phase === 'start' } as AgentRetryDiagnostic
  } else if (event.type === 'user_message') {
    state.message_id = event.correlation_id
    state.retry = null
  }
  return state
}
